### PR TITLE
Escape fewer Unicode codepoints in `Debug` impl of `str`

### LINKED ISSUE #34318: Unreadable io::Error debug strings
While trying to open a non-existing file (on Windows 10, Chinese):

```rust
use std::fs::File;
fn main() {
    File::open("no-such-file.rs").unwrap();
}
```

It panicked with a message:

```
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 2, message: "\u{7cfb}\u{7edf}\u{627e}\u{4e0d}\u{5230}\u{6307}\u{5b9a}\u{7684}\u{6587}\u{4ef6}\u{3002}" } }', ../src/libcore\result.rs:746
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

The OS native error message here is **unreadable** for human being.

The message (`"\u{7cfb}\u{7edf}\u{627e}\u{4e0d}\u{5230}..."`) is the _escaped_ result of "No such file or directory" in Chinese ("系统找不到指定的文件")。[`impl fmt::Debug for str`](http://doc.rust-lang.org/stable/src/core/up/src/libcore/fmt/mod.rs.html#1350-1368) do this escape.

I can read Chinese, which is my mother tongue, but I can't read `\u{7cfb}...`.

Possible solutions:
- Don't use local language (Chinese here) OS error string, use English instead.
- Change `impl fmt::Debug for str` to not escape most Unicode chars. (breaking-change?)

---

```
fn main() {
    let s = "Hello δεΣ❤ 日月";
    println!("{}", s);   // prints "Hello δεΣ❤ 日月"
    println!("{:?}", s); // prints "Hello \u{3b4}\u{3b5}\u{3a3}\u{2764} \u{65e5}\u{6708}"
}
```

[Run it on play.rust-lang.org](https://is.gd/T7lmqG)

Is `impl fmt::Debug for str` really friendly enough for _debug_ purpose? Is there possibility that we change its implementation as similar as `impl fmt::Display for str`?

### PR DESCRIPTION
**Pull request by tbu-**
Use the same procedure as Python to determine whether a character is
printable, described in [PEP 3138](https://www.python.org/dev/peps/pep-3138/). In particular, this means that the
following character classes are escaped:
- Cc (Other, Control)
- Cf (Other, Format)
- Cs (Other, Surrogate), even though they can't appear in Rust strings
- Co (Other, Private Use)
- Cn (Other, Not Assigned)
- Zl (Separator, Line)
- Zp (Separator, Paragraph)
- Zs (Separator, Space), except for the ASCII space `' '` `0x20`

This allows for user-friendly inspection of strings that are not
English (e.g. compare `"\u{e9}\u{e8}\u{ea}"` to `"éèê"`).

Fixes #34318.
CC #34422.

### CODE REVIEW

**Comment by sfackler** [comment]
Looks like run-pass/ifmt.rs is failing on travis.

**Comment by ollie27** [comment]
Is this changing [`char::escape_default`](https://doc.rust-lang.org/nightly/std/primitive.char.html#method.escape_default)? That would be a breaking change.

**Comment by tbu-** [comment]
It is changing that function. Why is it a breaking change?

**Comment by ollie27** [comment]
It's a stable function and this will break people's code which relies on the current behaviour.

**Comment by ranma42** [comment]
The [documentation](http://doc.rust-lang.org/std/primitive.char.html#method.escape_default) explicitly states that any character that is not in the *printable ASCII* range `0x20 .. 0x7e` will be escaped (some characters with ad-hoc sequences, the other ones with hexadecimal Unicode escapes), so this change would at the very least need to change the documentation of the function to match the new behaviour.
This is probably undesirable as the API is stable and the current specification guarantees that the output is only composed of ASCII characters, while the new implementation has much looser and complex guarantees (for example, it would be hard to provide an accurate specification without listing the Unicode character classes that are not escaped).

**Comment by BurntSushi** [comment]
@ranma42 Agreed. I don't think we can change this. It's very clearly changing the contract of the function.

**Comment by tbu-** [comment]
@ranma42 OK, assume for now that we don't change that function, but only the `Debug` implementation of `str`.

**Comment by eddyb** [comment]
Didn't these exact conversations happen before? Was a previous attempt abandoned?

**Comment by tbu-** [comment]
I don't know, I don't remember any.

**Comment by eddyb** [comment]
I can't find it. Might have been `char` instead. Don't mind me.

**Comment by ranma42** [comment]
@tbu- , yes, I think that should work. We might also want to expose it as a function on `char` (`escape_nonprintable`?), but I am unsure if this should be done directly or if we should go for an RFC to discuss the details (which `Debug` impls should escape non printable chars and which ones should escape everything? should we expose the API on char? under what name? are there other conventions besides the one established by Python that might be worth evaluating?).

**Comment by brson** [comment]
Not sure what I think of this.

The medium that `Debug` writes to is a byte sequence with no guarantees that it supports Unicode. Certainly most places the output would end up would speak UTF-8 but definitely not all.

**Comment by tbu-** [comment]
Actually, the output of `Debug` is statically guaranteed to be UTF-8.

I would *guess* that the reason for this is that the people implementing it didn't need non-ASCII characters, and I mean if you don't need them they're just a nuisance. But if you're implementing a non-English program, then it basically makes the `Debug` implementation useless (see the linked issue in which @liigo shows the unreadable "file not found" message in Chinese.

If you write to a device that doesn't support UTF-8, you should just escape these characters later, when writing to said device -- like the `ascii` function in Python. The other way around doesn't work.

**Comment by ranma42** [comment]
> But if you're implementing a non-English program, then it basically makes the Debug implementation useless

A possible objection is that

> `Debug` should format the output in a programmer-facing, debugging context.

which seems to imply that it should not be exposed to the users, but rather to tools or developers.
For example, one could say that formatting errors using `Debug` is not correct if they are going to be shown to the user.

**Comment by tbu-** [comment]
@ranma42 The linked example for a programmer, @liigo.

**Comment by ranma42** [comment]
@tbu- not really, in that case he is the user of the `rustc` program.

**Comment by tbu-** [comment]
@ranma42 It's a runtime error provided by the operating system, encountered while programming.

**Comment by liigo** [comment]
> `Debug` should format the output in a programmer-facing, debugging context.

I'd rather they *don't* do these (mostly useless) format/escape for me (a programmer). They ~~(self-righteous people)~~ *thought* these format may help programmers in debugging context, but it's not true. In some times, non-English programmers got [unreadable](https://github.com/rust-lang/rust/issues/34318) formatted output, **worse** than `Display` output. I don't use *very-old* output devices, which maybe have problems to display Unicode characters, which really need escape.

@tbu- Maybe we can't change `escape_default` for compatible reasons, but changing `impl Debug for str` is just implementation details IMHO.

**Comment by ranma42** [comment]
@tbu- It is a runtime error provided by the operating system, encountered by `rustc` and dumped on the console in a bad way. What I am saying is that it is not obvious to me that the correct way to display it is using `Debug` instead of `Display`.

**Comment by liigo** [comment]
> @brson: it seems like the possibility of terminals not understanding it must be one reason.

If this was the reason, we should also implement `println!` to escape Unicode characters.

**Comment by liigo** [comment]
> @ranma42: to display it is using `Debug` instead of `Display`

That is a big (breaking?) change.

**Comment by ranma42** [comment]
@liigo Yes, that would be a major breaking change (it would change the constraints on the `Result` type so that `E: fmt:Display`).

**Comment by ranma42** [comment]
To me, the major advantage of the current implementation of `Debug` for `str` (always escaping non-ASCII) is that it allows to distinguish between different strings that have the same visual appearance. This applies in the same way to whitespace ('\t' vs spaces), Unicode combining characters and control characters... and maybe more?

Of course this does not mean that it should be used for everything. Specifically, I would only use `Debug` when the *representation* of the data is relevant, not when it just needs to be shown to the user. Basically this reasoning is my understanding of the documentation of `Debug` and `Display` (programmer-facing vs user-facing). This can of course be incorrect, but in that case I would also ask for an improvement in the documentation ;)

#34318 shows an example where using `Debug` looks inappropriate (an OS-provided localised error). A partial solution that would have limited impact on other things would be to avoid escaping *only* the OS localised errors [here](https://github.com/rust-lang/rust/blob/master/src/libstd/io/error.rs#L285). I would not even call this a breaking change, given that the output is already not explicitly stabilised (it obviously depends on the current locale and most likely on the operating system + version).

Even though Rust does not (yet) have its own localised error messages, it would not be hard to imagine the same issue affecting other types of output, so it might be a good idea to think of a more general solution to ensure a way forward in this direction.

**Comment by tbu-** [comment]
@ranma42 If you want to see the exact code points, why only make an exception for English? That's very English-centric. :)

We should probably provide a function that does the same as `Debug` today, like Python has: `ascii`. This basically solves the problem. If you need a `Debug` string like today, you can just put the `Debug` output into said function and you receive what you wanted.

**Comment by liigo** [comment]
> To me, the major advantage of the current implementation of Debug for str (always escaping non-ASCII) is that it allows to distinguish between different strings that have the same visual appearance.

This is not advantage for non-ASCII text. It just makes unreadable noise (`\u{xxxx}...`). If someone do think this is a major advantage, please let `println!` have this advantage too.

**Comment by bluss** [comment]
Current Debug is too english-centric, changing it makes development in other languages easier. Even when I write a program using ascii encoded source files only, it might handle non-ascii *data* (we support unicode strings after all), and escaping non-ascii in Debug is just another hurdle that makes development harder (Think of `#[derive(Debug)]` on a struct or larger data structure that contains strings).

**Comment by ranma42** [comment]
@tbu- You seem to be missing my point: if I use `Debug` I am not aiming at displaying the string, but rather its representation. For ASCII characters the representation is not ambiguous, thus very little escaping is needed. Instead it is basically impossible to distinguish some of the characters that I mentioned (whitespace, Unicode combining characters and control characters). For example when round-tripping strings across systems that perform different normalisations of Unicode strings it is possible to have unexpected results when comparing strings containing accented `è` and the combining accent followed by the `e` letter.

Of course one could say that instead of the `Debug` representation of the string, I should look at the associated `u8` slice to explicitly check the UTF8 encoding of the data if I need to check the representation. Nonetheless my objection so far has received no response: maybe `Debug` is not what you actually need. Why are we using `Debug` instead of `Display` for these strings?
From a conceptual point of view (i.e. besides implementation and backward compatibility issues) which one would be appropriate for this use?

@liigo I do think that is an advantage and `println!` already has it: `{}` can be used to print Unicode strings and `{:?}` to escape them.

**Comment by liigo** [comment]
so `{}` have less advantage than `{:?}`? I don't think so.

**Comment by liigo** [comment]
> @ranma42: maybe Debug is not what you actually need. Why are we using Debug instead of Display for these strings?

`Debug` is designed/used to *display* something, for *debug* purpose. `Result::unwrap()`s do use `Debug` the right way. And changing that would be a major breaking change.

**Comment by ollie27** [comment]
There is a trade-off here. When debugging code with non-ASCII strings, is it more important to be able to disambiguate all similar looking strings or be able to read the data stored in the string?

**Comment by ranma42** [comment]
> so `{}` have less advantage than `{:?}`? I don't think so.

I am unable to understand what you are trying to say here. The fact that `println!` has both does not speak in favour or against any of them. They are both available and each one is useful for different purposes. In fact, I am trying to argue that being different makes them *more useful* (If `Debug` and `Display` always produced the same result when both are available and `Debug` was implemented for more types than `Display`, then you could very reasonably say that one of them is more useful than the other).

> `Debug` is designed/used to display something, for debug purpose. `Result::unwrap()`s do use Debug the right way. And changing that would be a major breaking change.

The question is what "debug purpose" means. Apparently it can have multiple meanings: in some cases when debugging you might need to know if "caffè" is 5 or 6 chars (and which ones); in other cases hiding this representation detail and showing the "è" character is better.

Example: I would find it surprising to see a `HashSet` with the "same" string multiple times and to debug it I would probably prefer to see the escaped representation (current `Debug` implementation) to easily detect the issue. When I do not need to check this details, I would probably prefer a more readable representation. In any case, I would certainly not want to show an escaped version to the user.

@ollie27 I think both are important use cases. I am afraid that the standard library focuses too much on the first one, but I would rather try to make both more useable than just drop one in favour of the other one.

What if in addition to the characters suggested by the OP we escaped combining characters?
This would still allow to distinguish between different strings, but it would also allow more characters to go through unescaped.

Example: "caffè" could be escaped to "caffè" or "caffe\u{300}" (depending on its representation) but never to "caff\u{e8}".

Would this solve #34318 ? (I think so, at least none of the characters in the error seem to be a combining one)

**Comment by ollie27** [comment]
It's more than just combining characters though, there are plenty of individual characters which look the same.

As far as `io::Error` is concerned the `Debug` implementation probably shouldn't escape the string at all given that it shouldn't contain anything which needs escaping and there is no need for disambiguation.

**Comment by ranma42** [comment]
@ollie27 You are right, there are several characters with the same aspect :(

I think we all agree that for `io:Error` the OS-provided string should always be displayed as-is, with no escaping (not even of '\' and other ASCII characters).

**Comment by tbu-** [comment]
@ranma42 I think it should be escaped like strings would be after this pull request. Just so you can easily copy-paste it into source code.

**Comment by rkruppe** [comment]
It is true that debugging will occasionally require comparing code points, but I can't bring myself to believe that it's economical to optimize for that case. Many, many strings wind up being formatted via `Debug` simply because that is the only or easiest option: `Result::Err` is only one example, there's also

- other generic APIs which use `Debug` rather than `Display` as bound
- all the collections in std (how often have you dumped a `Vec<String>` during debugging?)
- Other types that contain (directly or as a type parameter) strings which don't implement `Display`, either because the author didn't think of it or consciously because there is no sensible user-facing representation of the type

Save for the simplest possible case (a plain `String`, or a small set of well-behaved wrapper types), you're much more likely to be forced into `Debug` than have the option to use `Display` on any given string. With types that you control, you theoretically have the option to implement `Debug` manually in a way that doesn't escape the strings, but this is incredibly bothersome and error prone compared to `derive(Debug)`, and when it's a third party type you're out of luck.

All these obstacles to just reading the effing string, for what? Just so the couple of times someone has a nasty encoding related problem they can find it more easily? Don't get me wrong, these kinds of bugs are extremely annoying and I've had to debug one or two myself. But this benefit stands in no relation to the problems it causes in the *vast majority of cases* when it's just a harmless string and you don't need to care the least about the number and numeric value of the code points.

I'm German, so 90% of text in my native language is just ASCII, but put yourself in the position of someone dealing with text that isn't mostly ASCII: Every time that the path of least resistance leads to some string being formatted with `Debug` (quite common, see above), it becomes **completely unreadable**. You can't tell me that doesn't make debugging harder as well.

**Comment by ranma42** [comment]
@tbu- if you want to copy-paste Unicode text in a robust way, you need to keep into account that it might be affected by normalisation (and it might not even always happen in the same way depending on .

@rkruppe the RFC that proposed the `Debug` and `Display` traits suggests that `Result::unwrap` should depend on `Display` [here](https://github.com/rust-lang/rfcs/blob/master/text/0565-show-string-guidelines.md#use-in-errors). Apparently this section was not implemented because of the post-merge discussion. Unfortunately localisation issues were not mentioned in that occasion, but it might make sense to discuss that again and reach a resolution that also amends the RFC text (ideally specifying in detail what is the meaning of "Strings print as normal quoted notation, with escapes.").

**Comment by rkruppe** [comment]
@tbu- As I said, `Result::unwrap` is just of many, many ways a string might get formatted through `Debug` even though there is no reason to perform escaping on it. `Debug` is so ubiquitous, and `Display` so comparatively specific and rare, that `Debug` is *not* an optional thing you can just turn on when you suspect an encoding error, but rather the path of least resistance for almost all developer-oriented output.

It's impossible to fix all or even most of those factors that force `Debug` on everything &mdash; even switching `Result::unwrap` from `Debug` to `Display` is a breaking change. It's not even clear whether that would be desirable. For example, it's intentional and probably good that `Vec` and other collections implement `Debug` but not `Display`, yet in conjunction with `<str as Debug>` being how it is today, this makes debugging output for text collections unreadable unless you happen to be one of those lucky people who only ever handle ASCII. (I should stress again that this is just a single example for concreteness, a band-aid for the specific case of `Vec<String>` doesn't help at all.)

The solution, then, is **not** to avoid `<str as Debug>` in a few locations where it causes the most obvious pains. The real solution is to make `<str as Debug>` more suitable for the average case, which is *human text*, not *containers of code points*.

**Comment by rkruppe** [comment]
Oops, @ranma42, not @tbu-

**Comment by brson** [comment]
Triage: still waiting for @rust-lang/libs on this.

I'm inclined to make `Debug` print more readable text.

**Comment by alexcrichton** [comment]
Thanks again for the PR @tbu-! The libs team got a chance to discuss this today and the conclusion was that we would like to merge this, but we shouldn't change the behavior of the existing `escape_default` method. We can probably add a *new* method corresponding to this new behavior, and then we can document it as "this is how Rust escapes strings" without strictly documenting what's escaped.

**Comment by retep998** [comment]
I'd personally prefer just making `io::Error` display the error message without escaping, without modifying escaping elsewhere.

**Comment by liigo** [comment]
@retep998 But how to do it? Not easy.

**Comment by retep998** [comment]
@liigo Just write a simple wrapper type which wraps a `&str` with a manual `Debug` impl that prints the quotes but doesn't do any escaping. Then wrap the error string in that wrapper type [here](https://github.com/rust-lang/rust/blob/e7751e436ba4ff05dd844e330ee50afb7123e1fa/src/libstd/io/error.rs#L311).

**Comment by alexcrichton** [comment]
ping @tbu-, any update on this PR?

**Comment by tbu-** [comment]
Working on it.

**Comment by tbu-** [comment]
@alexcrichton Do you know whether the Travis error is my fault?

**Comment by tbu-** [comment]
Unwind error 3 seems to be `_URC_FATAL_PHASE1_ERROR`.

**Comment by TimNN** [comment]
@tbu- The travis failure is a known issue (so not related to this pr) and a fix should be with @bors at the moment.

**Comment by tbu-** [comment]
@alexcrichton Except for the failure that Travis gets and the unrelated one I experience locally, this should be good to go.

File: src/libcore/char.rs [hunk 1]
```diff
@@ -263,6 +264,8 @@ pub trait CharExt {
     fn escape_unicode(self) -> EscapeUnicode;
     #[stable(feature = "core", since = "1.6.0")]
     fn escape_default(self) -> EscapeDefault;
+    #[unstable(feature = "char_escape", issue = "0")]
```
Comment (alexcrichton) [inline src/libcore/char.rs]: This should get filled in with a tracking issue, and could you open up a new issue in lieu of #27791 just to refresh the discussion a bit?

**Comment by alexcrichton** [comment]
Thanks @tbu-! cc @rust-lang/libs, thoughts about the name `str::escape`, `char::{escape, Escape}`?

**Comment by brson** [comment]
It's a pretty sweet name, one that we didn't take before - and in retrospect it was good we didn't take it because we made mistakes. Why do we think this function is *the one* to get the best name? It's not even on a path to stability. I'd suggest `escape_debug` to reflect it's purpose.

**Comment by tbu-** [comment]
@alexcrichton Added tracking issue, renamed to `escape_debug`.

File: src/libcollections/str.rs [hunk 2]
```diff
@@ -1697,6 +1697,14 @@ impl str {
         return s;
     }
 
+    /// Escapes each char in `s` with `char::escape`.
+    #[unstable(feature = "str_escape",
+               reason = "return type may change to be an iterator",
+               issue = "27791")]
+    pub fn escape(&self) -> String {
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: This may want to be called `escape_debug` as well

File: src/libcore/fmt/mod.rs [hunk 3]
```diff
@@ -1383,7 +1383,7 @@ impl Debug for str {
         f.write_char('"')?;
         let mut from = 0;
         for (i, c) in self.char_indices() {
-            let esc = c.escape_default();
+            let esc = c.escape();
```
Comment (alexcrichton) [inline src/libcore/fmt/mod.rs]: These should be updated as well I think?

(may be good to do a smoke test with `make check`

**Comment by brson** [comment]
@borss r+ Thanks @tbu- .

**Comment by tbu-** [comment]
@brson Could you @bors r-, there's another test failure in the WTF-8 code.

**Comment by tbu-** [comment]
@alexcrichton It should be fixed now.

**Comment by SimonSapin** [comment]
I’m very late to say this, but this adds 2102 bytes of static data to libcore, whereas previously all large Unicode tables were in the `std_unicode` crate that `#![no_std]` programs could opt not to use. 2 KB may not seem like much, but it’s significant when programming a micro-controller that has 16 KB of flash memory.

**Comment by aturon** [comment]
@SimonSapin I opened https://github.com/rust-lang/rust/issues/39492 for this issue, and to propose a general policy.

**Comment by ariasuni** [comment]
I’d like to know what the code do precisely (what are `SINGLETONS0U`, `SINGLETONS0L`, `NORMAL0` and `check()`) so I could try to optimize it, at least in size. Also, this code should to belong to `libstd_unicode`, right?

**Comment by SimonSapin** [comment]
@ariasuni The commit message and PR message give a list of Unicode categories of characters that are escape, but yes if it’s not already this list should also be in some doc-comment in the code.

I agree that I’d prefer to have these tables in `libstd_unicode`, but `libcore` still needs *some* impl of `Debug for str`. There can only be one such impl in a program, I don’t know if it’s possible to make it do something different based on whether `libstd_unicode` is available.

**Comment by tbu-** [comment]
@ariasuni The high-level view is that you have to store `0x110000` booleans, one for each Unicode code point, which specify whether the given code point should be printed or not (according to the description in the first post of this PR).

The low-level view of this particular implementation seems to have changed since I have implemented it, you can find some notes on the new one in 44bcd261a7fa2a4ad873a77b67ac88b0cf09f111.

**Comment by ariasuni** [comment]
@SimonSapin In fact, we can split `libstd_unicode` in `unicode-casing` and `unicode-categories`, and use only the latter to implement `is_printable()`. What do you think about it?

**Comment by SimonSapin** [comment]
@ariasuni even if we did that, that doesn’t solve the situation that a `#[no_std]` program needs a `Debug for str` impl and ideally should also be able to opt out of embedding large Unicode tables. If we move `is_printable` to a separate crate but make `libcore` depend on that crate, we’ve moved things around without actually changing anything for Rust users. If we add a Cargo feature to disable these including tables (https://github.com/rust-lang/rust/issues/39492), it doesn’t change much that it’s a separate crate or not.

**Comment by ariasuni** [comment]
Sorry, I probably overthought it. The idea is to put the code for Unicode categories elsewhere so that we can use it in `is_printable()`. This would reduce the code size for those who don’t use `#[no_std]`.

**Comment by SimonSapin** [comment]
I don’t understand how this would reduce code size, I may be missing something.

**Comment by ariasuni** [comment]
If I understand correctly, `is_printable()` use Unicode categories data in `char_private.rs`, instead of using the data in `libstd_unicode` used by `is_control()` `is_alphabetic()`, `is_numeric()`, etc.

**Pull request merged**